File: loomoroutines-dispatcher/src/main/java/dev/reformator/loomoroutines/dispatcher/Dispatcher.java
package dev.reformator.loomoroutines.dispatcher;

import org.jetbrains.annotations.NotNull;

import java.time.Duration;

public interface Dispatcher {
    default void execute(@NotNull Runnable action) {
        scheduleExecute(Duration.ZERO, action);
    }

    void scheduleExecute(@NotNull Duration duration, @NotNull Runnable action);

    default boolean canExecuteInCurrentThread() {
        return false;
    }

//    Dispatcher VirtualThread = new Dispatcher() {
//        @Override
//        public void execute(@NotNull Runnable action) {
//            Thread.ofVirtual().start(action);
//        }
//
//        @Override
//
//        public void scheduleExecute(@NotNull Duration duration, @NotNull Runnable action) {
//            Thread.ofVirtual().start(() -> {
//                try {
//                    Thread.sleep(duration);
//                } catch (InterruptedException e) {
//                    throw CommonUtils.throwUnchecked(e);
//                }
//                action.run();
//            });
//        }
//
//        @Override
//        public boolean doDispatchInCurrentThread() {
//            return Thread.currentThread().isVirtual();
//        }
//    };
}

File: loomoroutines-dispatcher/src/main/kotlin/dispatcher.kt
package dev.reformator.loomoroutines.dispatcher

import dev.reformator.loomoroutines.common.internal.Action
import dev.reformator.loomoroutines.common.internal.Callback
import java.time.Duration

interface Dispatcher {
    fun execute(action: Action) {
        scheduleExecute(Duration.ZERO, action)
    }

    fun scheduleExecute(delay: Duration, action: Action)

    fun canExecuteInCurrentThread(): Boolean =
        false
}

interface Promise<out T> {
    val state: PromiseState

    fun join(): T

    fun subscribe(callback: Callback<PromiseResult<T>>)
}

enum class PromiseState {
    RUNNING, COMPLETED, EXCEPTIONAL
}

sealed interface PromiseResult<out T> {
    fun get(): T

    val succeed: Boolean
}

class SucceedPromiseResult<out T>(private val result: T): PromiseResult<T> {
    override fun get(): T =
        result

    override val succeed: Boolean
        get() = true
}

class ExceptionalPromiseResult<out T>(private val exception: Throwable): PromiseResult<T> {
    override fun get(): T {
        throw exception
    }

    override val succeed: Boolean
        get() = false
}
